src/RelayClient.ts

```diff
@@ -103,6 +105,45 @@ class RelayClient extends EnvelopingEventEmitter {
       metadata,
     };
   };
+
+  public estimateInternalCallGas = async (
+    { request, relayData }: EnvelopingTypes.RelayRequestStruct,
+    internalGasCorrection: BigNumberish,
+    addCorrection = true
+  ): Promise<BigNumber> => {
+    let estimation = await this._provider.estimateGas({
+      from: request.from,
+      to: request.to,
+      gasPrice: relayData.gasPrice,
+      data: request.data,
+    });
+
+    // The INTERNAL_TRANSACTION_ESTIMATE_CORRECTION is substracted because the estimation is done using web3.eth.estimateGas which
+    // estimates the call as if it where an external call, and in our case it will be called internally (it's not the same cost).
+    // Because of this, the estimated maxPossibleGas in the server (which estimates the whole transaction) might not be enough to successfully pass
+    // the following verification made in the SmartWallet:
+    // require(gasleft() > req.gas, "Not enough gas left"). This is done right before calling the destination internally
+    if (estimation.gt(internalGasCorrection)) {
+      estimation = estimation.sub(internalGasCorrection);
+    }
+
+    if (addCorrection) {
+      estimation = this._applyGasCorrectionFactor(estimation);
+    }
+
+    return estimation;
+  };
+
+  private _applyGasCorrectionFactor(estimation: BigNumberish): BigNumber {
+    let bigEstimation = BigNumberJs(estimation.toString());
+    const bigGasCorrection = BigNumberJs(ESTIMATED_GAS_CORRECTION_FACTOR);
```

Comment:
I don't like the fact that here we're using this constant without having the possibility to configure it.
I think it would more flexible if we had an optional parameter `esimatedGasCorrectFactor` that if not specified uses the constant.
We should probably specify that parameter in the method `estimateInternalCallGas` so that the `_applyGasCorrectionFactor` can receive it from there.

Reply:
updated